### DISCUSSION TITLE
how to graceful shutdown the kafka cluster ?

### DISCUSSION THREAD

**Discussion by ahu0605** [strimzi/strimzi-kafka-operator#6837] [Q&A]
I know kafka cluster had graceful shutdown process, leave clutser and sychronize data,election leader? when strimzi kafka will shutdown,how to kill pod? kubenetes will asychronized resource? can you tell me the code how to implements? I have no idea

**Comment by scholzj** [comment]
Maybe I don't understand what re you trying to do. But as long as the termination grace period is long enough, the pod just shutdowns cleanly when you delete it (which basically sends it SIGTERM signal).

**Reply by ahu0605** [reply]
your means add the shutdown secords？

**Reply by scholzj** [reply]
The termination grace period is a Pod option.

**Reply by ahu0605** [reply]
> The termination grace period is a Pod option.

"controlled.shutdown.enable" do you know the config item? maybe shutdown will fast or long long time...because data will flush disk and election leader, do you mean "PreStop" hook?

**Reply by scholzj** [reply]
Sorry, I don't really get what you are asking for or what point are you trying to make. `controlled.shutdown.enable` is enabled by default and the termination grace period defines how much time it gets to shutdown cleanly. There is no pre-stop hook needed for this.

**Reply by ahu0605** [reply]
> Sorry, I don't really get what you are asking for or what point are you trying to make. `controlled.shutdown.enable` is enabled by default and the termination grace period defines how much time it gets to shutdown cleanly. There is no pre-stop hook needed for this.

when broker node is busy , I think kafka shutdown need long time ,the pod will be killed ,maybe not clean shutdown

**Reply by scholzj** [reply]
Well, there is a trade-off. You can set the grace period to something very high to protect against situations like this. But then the termination might take a long time even when the pod is dead and you would want to roll it. In any case, the containers need to be able to terminate sooner or later.

**Reply by ahu0605** [reply]
our test man，discovery when the kafka node be killed un graceful， when the node start agai， the topic leader not rebalance！ but when you exec shudown.sh wait until gracefule shutdown， the kafka will election the leader for balance， maybe kafka need the kubernetes wait long time...

**Reply by scholzj** [reply]
When you kill Kafka in _ungraceful_ way (e.g. by force deleting the pod), you kill it in ungraceful way. There is not much anyone can do with it. And it can happen anytime when for example your worker node dies. For a clean shutdown, it is about the time you give it. And that is where - as I said - you can use and adjust the termination grace period.

**Reply by ahu0605** [reply]
> When you kill Kafka in _ungraceful_ way (e.g. by force deleting the pod), you kill it in ungraceful way. There is not much anyone can do with it. And it can happen anytime when for example your worker node dies. For a clean shutdown, it is about the time you give it. And that is where - as I said - you can use and adjust the termination grace period.

yes，I know，if busy node  check unclean shutdown process, will recovery log segement and index file, it is spent long time,the kubernetes will health check and failed .
